### ISSUE TITLE
Comments or Intuition why these tricks work?

### ISSUE THREAD

**Issue by spurra** [soumith/ganhacks#3]
Hi,
Thanks for the interesting collection. Is it possible to add a comment to each trick on why it actually helps training?
Thanks!

**Comment by ZhimingZhou** [comment]
Similar issue here. Could you explain more, the reason behind?

For example, I'm wondering that 

1. What'll happen when label smoothed to a value large than 1.0? and why does here suggest negative label smooth also? Seems not consistent with the paper.

2. When sample z from Gaussian, what if you get a very far away sample? that could happen, though rare. Should we instead sample from a high dimensional circle?

3. I'm not sure, but the "for example" in " balance loss " means it is a good example, right? It seem good to me. 

Many Thanks!

**Comment by soumith** [comment]
most of these tricks are empirically found. it's not possible to add a comment on why each trick actually helps training without having the possibility of being very very wrong and leading folks the wrong way.

> When sample z from Gaussian, what if you get a very far away sample? That could happen, though rare. Should we instead sample from a high dimensional sphere?

This is unlikely, almost all the mass on high dimensional sphere is on the sphere surface. that being said, i think it'll be okay to just sample from the surface of a high dimensional sphere.

> I'm not sure, but the "for example" in " balance loss " means it is a good example, right? It seem good to me.

Do you mean the "while" example? It does not work.

**Comment by ZhimingZhou** [comment]
@soumith Thanks for the replies.

Could I consult one more question that what is the equivalent dimension of the surface of a high dimensional sphere? 

I just did some experiments: generate random samples from high dimensional sphere surface (via normalize high dimensional Gaussian noises), the result images are not bad, but show very similar appearance. Do you have any idea on why this happens? 

Thanks.

**Comment by spurra** [comment]
@ZhimingZhou This is a known problem in GANs and are called generator collapse. The generators distribution collapses on a few modes of the data-generating distribution and remains there. There are approaches proposed on how to deal with. A recent one, the papers name escapes me, deals with it by training the generator on discriminators of several epochs.

@soumith If I remember correctly, some of your tricks have a basic intuition on why they work. It would be immensly helpful to have these included. If you're worried that you may lead people in the wrong direction, you could add a disclaimer.

**Comment by ZhimingZhou** [comment]
@spurra Thanks for your comments. Yeah, it should be generator collapse. My concerning is why normalizing latent code (to make it on unit-sphere's surface) leads to the collapse. Sure, it could just happens when I did any change. Maybe I should test it with more hyper-parameters. 

By the way, I googled "escapes me", but didn't find it. Could you share me a link? I'm recently interested in collapse avoiding. That should help. 

Thanks.

**Comment by spurra** [comment]
@ZhimingZhou You mean normalising the noise vector? I never heard of normalising latent codes. "The papers name escapes me" is a way of saying I do not remember the papers name :)

**Comment by ZhimingZhou** [comment]
@spurra Yeah, normalizing the noise vector.

**Comment by spurra** [comment]
@ZhimingZhou I found the paper https://pdfs.semanticscholar.org/6e24/9cd98d1c88d314d22d2d920aa5aa55a9620e.pdf

**Comment by ZhimingZhou** [comment]
@spurra Thanks!
